README.md
ZenPanel
========

See files for license


0.75/ChangeLog

sam. sept. 19 18:41:43 CEST 2009
- fixed segfault with bogus symlink to desktop file (thanks Til)

Mon Aug 24 22:05:26 CEST 2009
- fixed segfault

Thu Dec 25 19:43:01 CET 2008
- Localization of icon names

Mon Nov 10 16:10:47 CET 2008
- Added category detection
- Added custom banner, title, icon option
- Added localization support


Fri Dec 29 14:53:00 CET 2006
- Added command line arguments support for exec()
- Code cleanup
committer Jean-Philippe Guillemin <jp.guillemin@free.fr>


Mon Dec 18 09:45:34 CET 2006
- Added GTK icon theme support
committer Jean-Philippe Guillemin <jp.guillemin@free.fr>


Tue Dec 12 21:53:43 CET 2006
- Simple Makefile
- Simple Gtkdialog widget instead of menu 
- Banner support
- PATHS and DEBUG mode are DEFINED from Makefile
- Geometry CLI options
- Themed window title icon
- Simplified code
- Paths changes
- Commented out the scrolled box, useless since geometry can be specified.
committer Jean-Philippe Guillemin <jp.guillemin@free.fr>



commit e2b297407135b0ccf16bed9d485dbdf5ebdcf157
tree e30561bb0dfde814490c5e143d254ff40bb336c3
parent 0f018ed65e38eaeff59bfb977cbccaca1252909c
author Vincent Kergonna <vincent@phoebe.friends.paris> Fri, 03 Feb 2006 02:15:23 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Fri, 03 Feb 2006 02:15:23 +0100

    Changes the default directory for zenpanels ".desktop" files.

commit 0f018ed65e38eaeff59bfb977cbccaca1252909c
tree 7431aeb3205de6a0a088424dabe7ea1dfd1aeec3
parent 7ac3f084a9fb2e5de0677819b0002027d61e6615
author Vincent Kergonna <vincent@phoebe.friends.paris> Wed, 01 Feb 2006 23:45:36 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Wed, 01 Feb 2006 23:45:36 +0100

    Version 0.3

commit 7ac3f084a9fb2e5de0677819b0002027d61e6615
tree 8822170dd02e26b5e6e583722faa645e067f2452
parent abbd95ea13ed1d6b7ba32bcfa59c3bd9cf2da868
author Vincent Kergonna <vincent.k@orange.fr> Wed, 01 Feb 2006 22:42:59 +0100
committer Vincent Kergonna <vincent@joey.friends.paris> Wed, 01 Feb 2006 22:42:59 +0100

    Adds command line arguments parsing.

commit abbd95ea13ed1d6b7ba32bcfa59c3bd9cf2da868
tree e2490ba8b667f6c2707b612ce1cc49d1bbba1e5a
parent 190848a5de64b2d0adbbee07f86f7e4c79ef2933
author Vincent Kergonna <vincent.k@orange.fr> Wed, 01 Feb 2006 21:13:05 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Wed, 01 Feb 2006 21:13:05 +0100

    Icons are now created correctly in the panel after reading from .desktop
    files. Also almost every warning has been fixed.

commit 190848a5de64b2d0adbbee07f86f7e4c79ef2933
tree e33457148d52cb1015619aeedcbb6246b5a27d8d
parent 52d4ac9bcc020c7ff407f4249ab5b1edb16c2d6a
author Vincent Kergonna <vincent.k@orange.fr> Sun, 22 Jan 2006 23:36:00 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Sun, 22 Jan 2006 23:36:00 +0100

    We need to read only the .desktop files and no other files!

commit 52d4ac9bcc020c7ff407f4249ab5b1edb16c2d6a
tree 33f8f55aac1536fd848166908450401723b04905
parent f27b33c7801121210b7e9ab431e8d25615f06e00
author Vincent Kergonna <vincent@joey.friends.paris> Sun, 22 Jan 2006 19:36:38 +0100
committer Vincent Kergonna <vincent@joey.friends.paris> Sun, 22 Jan 2006 19:36:38 +0100

    This evolution now makes zenpanel reads it's .desktop files from a given
    directory instead of hardcoding everything.

commit f27b33c7801121210b7e9ab431e8d25615f06e00
tree 430b47b2ca3ba1974469620176b5d7a345b03bf1
parent 512a5b37e543794ae18f89268ab087f5d9f842cd
author Vincent Kergonna <vincent@joey.friends.paris> Sun, 22 Jan 2006 19:05:38 +0100
committer Vincent Kergonna <vincent@joey.friends.paris> Sun, 22 Jan 2006 19:05:38 +0100

    Adds files to allow parsing of .desktop files in zenpanel.

commit 512a5b37e543794ae18f89268ab087f5d9f842cd
tree c99d5b73bc86122737e0f02a1240533937261e0c
parent e78ce62889f956d77ea98556f3df4e959354638b
author Vincent Kergonna <vincent.k@orange.fr> Sat, 14 Jan 2006 00:47:09 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Sat, 14 Jan 2006 00:47:09 +0100

    Removes the doc dir from the main makefile. The doc dir will be created by
    the package build script so we don't need this anymore otherwise it will
    duplicates files.

commit e78ce62889f956d77ea98556f3df4e959354638b
tree d232d2e1a9ac7a2ce2ef025174690184280059ae
parent f73b543e4aa344bba08d6fac7aaa4861e0697b50
author Vincent Kergonna <vincent.k@orange.fr> Fri, 13 Jan 2006 21:01:34 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Fri, 13 Jan 2006 21:01:34 +0100

    Changed default window size. JP wants it this size so I've done it.

commit f73b543e4aa344bba08d6fac7aaa4861e0697b50
tree e5bf1d910dc655858dc2fe39b7ed2ff42a21b2b7
parent e24cce65a2f2df87bc7667bfd949f9e36093c70d
author Vincent Kergonna <vincent.k@orange.fr> Thu, 12 Jan 2006 23:50:11 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Thu, 12 Jan 2006 23:50:11 +0100

    Adds signal handling to know when a child is dead and finish it properly by
    calling wait4.

commit e24cce65a2f2df87bc7667bfd949f9e36093c70d
tree a62e29606e5ec0a59857e0a597baa3dde426083a
parent 8e4665c62805094359cc94c864c45ec460744808
author Vincent Kergonna <vincent.k@orange.fr> Thu, 12 Jan 2006 22:44:06 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Thu, 12 Jan 2006 22:44:06 +0100

    Resized window to 640x480 by default.
    Version 0.2

commit 8e4665c62805094359cc94c864c45ec460744808
tree 5e5cf5bb2051a8a1883d65debd8a4224bf644e62
parent 5bd26a2bf42110e0ef8e49c2a7143ae711a7b14a
author Vincent Kergonna <vincent.k@orange.fr> Mon, 19 Dec 2005 23:35:10 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Mon, 19 Dec 2005 23:35:10 +0100

    Add acconfig.h

commit 5bd26a2bf42110e0ef8e49c2a7143ae711a7b14a
tree 6bde38ad5eb8cfe9f6d56af74629dc59e28a82e9
parent 79bced6351cd2ca7823c93b4806cbfa76e73aa48
author Vincent Kergonna <vincent.k@orange.fr> Mon, 19 Dec 2005 23:32:37 +0100
committer Vincent Kergonna <vincent@phoebe.friends.paris> Mon, 19 Dec 2005 23:32:37 +0100


0.75/zenpanel.h
/*
 *  zenpanel.h
 *  Authors : 
 * - Vincent Kergonna <vincent.k@orange.fr> 
 * - Jean-Philippe Guillemin <jp.guillemin@free.fr>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdio.h>
#include <getopt.h>
#include <dirent.h>
#include <errno.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <strings.h>
#include <sys/utsname.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/resource.h>
#include <sys/wait.h>
#include <gtk/gtk.h>
#include <glib.h> 
#include <libintl.h>
#define _(String) gettext (String)

#ifdef DEBUG
#define dbg(fmt,args...) fprintf(stderr, fmt , ## args)
#else
#define dbg(fmt,args...) do { } while (0)
#endif

#define VERSION "0.75"

#ifndef _DEFAULT_DESKTOPPATH
#define _DEFAULT_DESKTOPPATH "/usr/share/applications"
#endif

#ifndef _DEFAULT_BANNER
#define _DEFAULT_BANNER "/usr/share/systemtools/banner.png"
#endif

#ifndef _DEFAULT_CATEGORY
#define _DEFAULT_CATEGORY "X-Zenpanel"
#endif

#ifndef _DEFAULT_TITLE
#define _DEFAULT_TITLE "Control Panel"
#endif

#ifndef _DEFAULT_ICON
#define _DEFAULT_ICON "zenpanel"
#endif

FILE * z_open_file (const char * filename);
char *z_read_parameter (const char *pname, FILE *fp);
void z_close_file (FILE *fp);
void  z_parse_cmd(char *line, char **argv);

typedef struct  {
	int toto;
	gchar filename[512];
	gchar icon[256];
	gchar name[256];
	gchar cmd[512];
	gchar categories[256];
}Z_desktop_file;



0.75/zpnl_parser.c
/*
 *  z_parser.c
 *  Authors : 
 * - Vincent Kergonna <vincent.k@orange.fr> 
 * - Jean-Philippe Guillemin <jp.guillemin@free.fr>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */


#include "zenpanel.h"

FILE * z_open_file (const char * filename){
	FILE *fp = NULL;
	fp = fopen (filename, "r");
	if (fp) return fp;	
	dbg("can't open file : %s\n", filename);
	return NULL;
}


char * z_read_parameter (const char *pname, FILE *fp){
	char buffer[512];
	char *value = NULL;
	char *sub=NULL;
	char *p=NULL;
	
	/* look for a given parameter in the .desktop file. */
	if (!fp) {
		return NULL;
	}
	
	memset (buffer, 511, 0);
	rewind (fp);
	do {
		p = fgets (buffer, 511, fp);
		if (p) {
			sub = strstr (buffer, pname);
		} else {
			break;
		}
	} while (!sub);
	
	if (sub) {
		p = strchr (buffer, '=');
		if (p) {
			p+=1;
			value = strdup (p);
		}
		return value;
	}else{
		return "not found";
	}
free(buffer);
}

/*
 * Closes the currently opened file pointed by fp
 */
void z_close_file (FILE *fp){
	fclose (fp);
	fp = NULL;
}

/*
 * Parses the cmdline into arguments
 */
void  z_parse_cmd (char *line, char **argv){
     while (*line != '\0') {       /* if not the end of line ....... */ 
          while (*line == ' ' || *line == '\t' || *line == '\n')
               *line++ = '\0';     /* replace white spaces with 0    */
          *argv++ = line;          /* save the argument position     */
          while (*line != '\0' && *line != ' ' && 
                 *line != '\t' && *line != '\n') 
               line++;             /* skip the argument until ...    */
     }
     *argv = '\0';                 /* mark the end of argument list  */
}


0.75/zenpanel.c
/*
 *  zenpanel.c
 *  Authors : 
 * - Vincent Kergonna <vincent.k@orange.fr> 
 * - Jean-Philippe Guillemin <jp.guillemin@free.fr>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */



#include "zenpanel.h"

/* usage */
int usage(void) {
	printf(_("\nzenpanel v%s : Zenwalk System Tools Manager\n"), VERSION);
	printf(_("Usage : zenpanel [options] ...\n"));
	printf(_("[-b banner] : path to the banner image\n"));
	printf(_("[-t title] : Window title\n"));
	printf(_("[-i icon] : icon name in theme\n"));
	printf(_("[-c Category] : Freedesktop Category that must match\n"));
	printf(_("[-d directory] : where to scan for desktop files\n"));
	printf(_("[-W] : window Width\n"));
	printf(_("[-H] : window Height\n"));
	return(EXIT_SUCCESS);
}

const gchar *desktop_dir = {_DEFAULT_DESKTOPPATH};
const gchar *category = {_DEFAULT_CATEGORY};
const gchar *banner = {_DEFAULT_BANNER};
const gchar *windowtitle = {_DEFAULT_TITLE};
const gchar *windowicon = {_DEFAULT_ICON}; 

gint width = 550;
gint height = 250;


enum {
	ICON_COLUMN,
	TEXT_COLUMN,
	DATA_COLUMN
};


static void destroy (GtkWidget *widget, gpointer data){
	gtk_main_quit();
}

void child_exited (pid_t pid)
{
	pid = waitpid(-1, NULL, WNOHANG);
}

static void icon_activated (GtkIconView *view, GtkTreePath *tree, gpointer data){
	pid_t pid;
	GtkListStore *store;
	GtkTreeIter iter;
	gchar *cmd = NULL;
	gchar *argv[64];
	store = GTK_LIST_STORE (data);

	gtk_tree_model_get_iter (GTK_TREE_MODEL (store), &iter, tree);
	gtk_tree_model_get (GTK_TREE_MODEL (store), &iter,
                     DATA_COLUMN, &cmd, -1);
	
	z_parse_cmd(cmd, argv); 
                     
	dbg("cmd=%s\n",cmd);
	pid = fork();
	if (pid == 0) {
		if (execvp(*argv, argv)) {
			exit(EXIT_FAILURE);
		}
	} else if (pid < 0) {
		dbg("Unable to fork");   
	} else if (pid > 0) {
		return;
	}
	return;
}


static void add_item_to_store (GtkListStore *store,
								gchar *icon_name,
								gchar *text,
								gchar *cmd){
					
	GtkTreeIter iter;
	GtkIconTheme *icon_theme;
	GdkPixbuf *pixbuf = NULL;
	
	if (!icon_name || !text || !cmd) {
		return;
	}
	
	icon_theme = gtk_icon_theme_get_default();	
	gtk_list_store_append (store, &iter);
	
	pixbuf = gtk_icon_theme_load_icon (icon_theme,
                                   icon_name, /* icon name */
                                   32, /* size */
                                   0,  /* flags */
                                   NULL);
	
	if (pixbuf) {
		gtk_list_store_set (store, &iter, ICON_COLUMN, pixbuf, 
			TEXT_COLUMN, text, DATA_COLUMN, cmd, -1);
		g_object_unref (pixbuf);
	} else {
		dbg("cannot create pixbuf from icon %s\n", icon_name);
	}
}

/*
 * This function reads .desktop files from a given directory and create the
 * icons in the panel.
 * 
 */
void
display_icons_in_panel (GtkListStore *store, const char * dir){

	Z_desktop_file df;
	
	const gchar *locale = setlocale(LC_CTYPE, "");
	gchar name_locale[128];
	g_snprintf(name_locale, sizeof(name_locale), "%s%s", "Name[", g_strndup (locale, 2));
	
	GDir *d = g_dir_open(dir, 0, 0);
    const gchar *file;
    FILE *fd;
	if (d) {
		while((file = g_dir_read_name(d))) {	
			g_snprintf(df.filename, sizeof(df.filename), "%s/%s", dir, file);
			
			if (g_strrstr_len (df.filename, strlen(df.filename), ".desktop")) {
				fd = z_open_file (df.filename);
				if ( !fd ) continue;
				g_snprintf(df.categories, sizeof(df.categories), "%s", z_read_parameter ("Categories", fd));
				g_strstrip (df.categories);
				if (strlen(df.categories) > 0 ) {
					if ( g_strrstr_len(df.categories, strlen(df.categories), category) ) {
						
						g_snprintf(df.icon, sizeof(df.icon), "%s", z_read_parameter ("Icon", fd));
						g_snprintf(df.name, sizeof(df.name), "%s", z_read_parameter (name_locale, fd));
						if (g_strrstr_len (df.name, strlen(df.name), "not found")) {g_snprintf(df.name, sizeof(df.name), "%s", z_read_parameter ("Name", fd));}

						g_snprintf(df.cmd, sizeof(df.cmd), "%s", z_read_parameter ("Exec", fd));
						g_strstrip (df.icon);
						g_strstrip (df.name);
						g_strstrip (df.cmd);
						add_item_to_store (store, df.icon, df.name, df.cmd);
					}
				}
				z_close_file (fd);
			}
		}
		g_dir_close(d);
	}
}


void create_main_window (void){	
	GtkWidget *window = NULL;
	GtkWidget *button = NULL;
	GtkWidget *icon_view = NULL;
	GtkListStore *store = NULL;
	GtkWidget *bannerwidget = NULL;
	GdkPixbuf *pxbuf;
	window = gtk_dialog_new();
	gtk_window_set_title (GTK_WINDOW (window), windowtitle);

	
	gtk_window_resize (GTK_WINDOW (window), width, height);
	gtk_window_set_icon_name (GTK_WINDOW (window), windowicon);
	
	
	/* The banner */
	pxbuf = gdk_pixbuf_new_from_file_at_scale(banner, width + 25, 35, FALSE, NULL);
	bannerwidget = gtk_image_new_from_pixbuf (pxbuf);
	
	g_signal_connect(G_OBJECT(window),
			 "delete_event", G_CALLBACK(destroy), NULL);

	g_signal_connect(G_OBJECT(window),
			 "destroy", G_CALLBACK(destroy), NULL);

	/* We create and define the icon view */
	icon_view = gtk_icon_view_new ();
	gtk_icon_view_set_item_width (GTK_ICON_VIEW (icon_view), 120);
	gtk_icon_view_set_row_spacing (GTK_ICON_VIEW (icon_view), 12);
	gtk_icon_view_set_column_spacing (GTK_ICON_VIEW (icon_view), 15);
	
	store = gtk_list_store_new (3, GDK_TYPE_PIXBUF, G_TYPE_STRING, G_TYPE_STRING);
	
	gtk_icon_view_set_pixbuf_column (GTK_ICON_VIEW (icon_view), ICON_COLUMN);
	gtk_icon_view_set_text_column (GTK_ICON_VIEW (icon_view), TEXT_COLUMN);
	gtk_icon_view_set_model (GTK_ICON_VIEW (icon_view), GTK_TREE_MODEL (store));
	
	g_signal_connect (G_OBJECT (icon_view), "item-activated", 
					G_CALLBACK(icon_activated), store);

	display_icons_in_panel (store, desktop_dir);
	
	button = gtk_button_new_from_stock("gtk-close");
	gtk_signal_connect (GTK_OBJECT (button), "clicked", G_CALLBACK (destroy), NULL);
	gtk_container_add (GTK_CONTAINER (GTK_DIALOG (window)->action_area), button);
	
	gtk_box_pack_start (GTK_BOX (GTK_DIALOG (window)->vbox), bannerwidget, FALSE, FALSE, 1);
	gtk_box_pack_start (GTK_BOX (GTK_DIALOG (window)->vbox), icon_view, TRUE, TRUE, 5);
	gtk_widget_show_all (window);
}


int main (int argc, char **argv){
	
	setlocale(LC_ALL, "");

	if(textdomain( "zenpanel" ) == NULL) {
		return EXIT_FAILURE;
	}

	if( bindtextdomain( "zenpanel", _LOCALEPATH ) == NULL) {
		return EXIT_FAILURE;
	}     
	textdomain ("zenpanel");

	/* I know it's archaic, as I am :)  */
	int i;	
	while ((i = getopt(argc, argv, "d:W:H:c:b:t:i:h")) != EOF){
		switch (i){
			case 'd':
				desktop_dir = optarg;
				break;
			case 'c':
				category = optarg;
				break;
			case 'b':
				banner = optarg;
				break;
			case 't':
				windowtitle = optarg;
				break;
			case 'i':
				windowicon = optarg;
				break;				
			case 'W':
				width = atoi(optarg);
				break;
			case 'H':
				height = atoi(optarg);
				break;
			case 'h':
				usage(); exit(EXIT_FAILURE);
		}
	}

	gtk_set_locale ();

	signal (SIGCHLD, child_exited);
		
	gtk_init (&argc, &argv);

	create_main_window ();

	gtk_main();
	
	exit (EXIT_SUCCESS);
}


